### ISSUE TITLE
New style multipage application refresh on a subpage routes back to default when authenticator is implemented

### ISSUE THREAD

**Issue by ShaunHide17** [mkhorasani/Streamlit-Authenticator#179]
Hi,

I am using the latest version of streamlit-authenticator (0.3.3) with the latest version of streamlit (1.37.0)

I am using the new multipage routing mechanism that streamlit recently released:
https://docs.streamlit.io/develop/concepts/multipage-apps/page-and-navigation

If I navigate away from the default page (View) to View2 then do a hard refresh (i.e. F5), it refreshes back to the default page (View) rather than the page I refreshed on. This behaviour does not occur without the `instantiate_authenticator()` function call so cant seem to work out what is driving the return to default behaviour

Can you offer any guidance?

Here is a sample of the main file from which the application is started with `streamlit run`

```
import streamlit as st
from helpers.authentication import instantiate_authenticator

st.set_page_config(page_title="Home", page_icon="", layout="wide")

instantiate_authenticator()
if st.session_state["authentication_status"]:

  view = st.Page("pages/view.py", title="View", default=True)
  view2 = st.Page("pages/view2.py", title="View2")

  navigation_dictionary = {
    "Views": [view, view2]
  }
  pg = st.navigation(navigation_dictionary)
  pg.run()
```
  
With the instantiate_authenticator() function being:

```
import os

import yaml
from yaml.loader import SafeLoader

import streamlit as st
import streamlit_authenticator as auth

def instantiate_authenticator():

    ## Load in the configuration from the environment variable
    config = yaml.load(os.environ["CONFIG"], Loader=SafeLoader)

    ## NOTE: Both the authenticator and the login must be triggered on every page.
    ##  Pushing the authenticator to st.session_state is to allow the user to logout

    ## Authenticator
    authenticator = auth.Authenticate(
        credentials=config["credentials"],
        cookie_name=config["cookie"]["name"],
        cookie_key=config["cookie"]["key"],
        cookie_expiry_days=config["cookie"]["expiry_days"],
        pre_authorized=config["pre-authorized"],
    )
    authenticator.login()

    if "authenticator" not in st.session_state:
        st.session_state.authenticator = authenticator

    if st.session_state["authentication_status"] == False:
        st.error("Username/password is incorrect")

```

**Comment by mkhorasani** [comment]
Hi @ShaunHide17, thank you for reaching out. Instead of using this custom function, can you try to add an `authenticator.login(location='unrendered')` widget to each other page to see if it will solve your problem?

**Comment by mkhorasani** [comment]
It seems that you are recreating several `stauth.Authenticate` objects, probably one for each page. If this is indeed the case, you will face a `DuplicateWidgetID` error, as the cookie manager does not use a unique key for each object. Instead, you should save the authenticator object to the session state once in your home page and access it from the session state from other pages.

**Comment by smboy** [comment]
can you help me understand a little bit here. when authenticator object is saved to the session state, and when user refreshes the page, wouldn't it loose the state? `st.session_state` is not persisted when page refreshes. So we loose the state and forces user to login. I'm having a similar issue in my app.

**Comment by mkhorasani** [comment]
Hi @smboy, session_state will only be lost if the user does a hard refresh. A normal refresh will not remove the session_state. But even if the user does for whatever reason do a hard refresh the authentication cookie will log them in automatically. I hope this answers your question.

**Comment by smboy** [comment]
Thanks for the details. Couple of followup questions:
- Just to clarify, this works seamlessly for a multipage application?
- also multi-user application without any issues?
- does this work across browsers too? seen case where Safari was not storing the cookie and when I refreshed (simple refresh) the page, it forced me to login.

Thanks again!

**Comment by mkhorasani** [comment]
> Thanks for the details. Couple of followup questions:
> 
> * Just to clarify, this works seamlessly for a multipage application?
> * also multi-user application without any issues?
> * does this work across browsers too? seen case where Safari was not storing the cookie and when I refreshed (simple refresh) the page, it forced me to login.
> 
> Thanks again!

Hi @smboy, can you verify that the issue you had was resolved after adding the `authenticate` object to session state? As for your questions, yes, yes, and yes. I've had no problems on my end.

**Comment by smboy** [comment]
It seems to be working on a standalone code. I will add it to my application and will update you (I was using a different authentication module which was having similar issues of persistence during page refresh).

meanwhile, anyway to customize the login? I would like to only display email and pwd..username is not required for my application. Thanks much again for your quick response!

**Comment by mkhorasani** [comment]
Yes, as of the latest version, users can now register their email as their username too. Feel free to change the name of the username field to email if you wish, you can do so using the `fields={'Username':'Email'}` argument.

**Comment by consoleLog7** [comment]
> Yes, as of the latest version, users can now register their email as their username too. Feel free to change the name of the username field to email if you wish, you can do so using the `fields={'Username':'Email'}` argument.

@mkhorasani this is a great new feature! I am trying to keep things simple for my users and just have email and password, so this helps.

I have this working on the login widget, however, can't get it working on register_user. When running `authenticator.register_user(pre_authorization=True, fields={'Username':'Email'})` I get errors "Missing submit button" and "DuplicateWidgetID" from line 347 in register_user `'Username' if 'Username' not in fields`. Not sure if this is jusr a me issue, or a feature request!

**Comment by mkhorasani** [comment]
> > Yes, as of the latest version, users can now register their email as their username too. Feel free to change the name of the username field to email if you wish, you can do so using the `fields={'Username':'Email'}` argument.
> 
> @mkhorasani this is a great new feature! I am trying to keep things simple for my users and just have email and password, so this helps.
> 
> I have this working on the login widget, however, can't get it working on register_user. When running `authenticator.register_user(pre_authorization=True, fields={'Username':'Email'})` I get errors "Missing submit button" and "DuplicateWidgetID" from line 347 in register_user `'Username' if 'Username' not in fields`. Not sure if this is jusr a me issue, or a feature request!

Since the `register_user` widget has both `username` and `email` fields in the form, if you rename the `username` field to `email` you will indeed get a DuplicatWidgetID error since both text fields now have the same name. I will try to modify this widget in the next release to remove the username field if the user wants to combine both fields into one. Please stay tuned!

**Comment by smboy** [comment]
If I may suggest, what would be appropriate is to take input fields as config parameter for registration and login. For instance, for registration, I want to use {user_email, pwd, company_name} and for login {username, password}. 

In fact, what I was thinking is to even isolate further. If there is any clean way where I can come up with my own login screen and simply pass the values to your application, that would be even ideal. That way, we are separating the front end with that of the backend. Its upto user on what and how they customize the login screen with.

**Comment by mkhorasani** [comment]
> If I may suggest, what would be appropriate is to take input fields as config parameter for registration and login. For instance, for registration, I want to use {user_email, pwd, company_name} and for login {username, password}.
> 
> In fact, what I was thinking is to even isolate further. If there is any clean way where I can come up with my own login screen and simply pass the values to your application, that would be even ideal. That way, we are separating the front end with that of the backend. Its upto user on what and how they customize the login screen with. Currently its tightly coupled with the login screen.

I definitely see a use case for this, but it would require major refactoring of the codebase and that is something that I can only commit to for the long term.

**Comment by christian-heins-velero** [comment]
i am still having the case where my page reloads to the default page, i have updated to the last version, provided authentication on the entry file, added the authenticator = st.session_state.authenticator to every page, my app still reloads to default or main page

**Comment by JurijsNazarovs** [comment]
I actulally don't understand how is it supposed to work. Can anyone provide an example of creating an authentification, and passing through multiple pages?

Assume i have page1.py
```python
# page1.py
from page2.py import render_page2

authenticator =  stauth.Authenticate(
        credentials=Config.auth_file,
        # Does providing config file only takes care of cookies names and values?
    )
 authenticator.login()
 render_page2(authenticator)
```

```python3
# Page 2
def render_page2(authenticator):
    authenticator.logout()
```

Is that a correctg way?

**Comment by Kaan191** [comment]
I had the same problem in my app and fixed this by creating the `st.navigation` object first, and then doing the auth flow check. 

Try changing the code from above as follows:

```python
import streamlit as st
from helpers.authentication import instantiate_authenticator

st.set_page_config(page_title="Home", page_icon="", layout="wide")

# define pages before authenticating
view = st.Page("pages/view.py", title="View", default=True)
view2 = st.Page("pages/view2.py", title="View2")

navigation_dictionary = {
    "Views": [view, view2]
}
pg = st.navigation(navigation_dictionary)

instantiate_authenticator()
if st.session_state["authentication_status"]:
  # ...but don't call `.run` until authenticated
  pg.run()
```
